src/WholeBodyStateDisplay.cpp

```diff
@@ -619,6 +679,43 @@ void WholeBodyStateDisplay::processWholeBodyState() {
       }
     }
 
+    // Center of pressure per contact. Mainly targets surface contacts (relatively meaningless for point contacts)
+    bool active_contact_in_foot_cop = false;
+    if (use_contact_status_in_foot_cop_) {
+      active_contact_in_foot_cop = contact.status == contact.ACTIVE;
+    } else {
+      active_contact_in_foot_cop = for_dir.norm() > force_threshold_;
+    }
+    bool is_contact_6d = false;
+    if (std::abs(contact.wrench.torque.x) > torque_threshold_ || std::abs(contact.wrench.torque.y) > torque_threshold_) {
+      is_contact_6d = true;
+    }
+    if (foot_cop_enable_ && active_contact_in_foot_cop && is_contact_6d) {
+      // NOTE: x component is negative due to right-hand rotation rule
+      Eigen::Vector3d foot_cop = Eigen::Vector3d( -contact.wrench.torque.y / contact.wrench.force.z,
+                                                     contact.wrench.torque.x / contact.wrench.force.z,
+                                                     0.0);  // Origin of frame is already at contact position
+      if (foot_cop_enable_ && std::isfinite(foot_cop(0)) && std::isfinite(foot_cop(1)) &&
+          std::isfinite(foot_cop(2))) {
+        Ogre::Vector3 foot_cop_pos(foot_cop(0), foot_cop(1), foot_cop(2));
+        foot_cop_pos_.push_back(foot_cop_pos);
+        boost::shared_ptr<PointVisual> foot_cop;
+        foot_cop.reset(new PointVisual(context_->getSceneManager(), scene_node_));
+        foot_cop->setPoint(foot_cop_pos);
+        foot_cop->setFramePosition(contact_pos);
+        foot_cop->setFrameOrientation(orientation);
+
+        // Setting color and properties
+        const float &radius = foot_cop_radius_property_->getFloat();
+        Ogre::ColourValue color = foot_cop_color_property_->getOgreColor();
+        color.a = foot_cop_alpha_property_->getFloat();
+        foot_cop->setColor(color.r, color.g, color.b, color.a);
+        foot_cop->setRadius(radius);
+
+        foot_cop_visual_.push_back(foot_cop);
+      }
+    }
+
```

Comment:
This is the core change. If contact torques are above a threshold (zero by default -- that may be a little lax), then we can choose to find a per foot, or per contact, centre of pressure.